File: 08. Trees/01. Binary Tree/16. Symmetric Tree/main.js
class TreeNode {
    constructor(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// DFS Approach
function isMirrorDFS(t1, t2) {
    if (!t1 && !t2) return true;
    if (!t1 || !t2) return false;
    return t1.val === t2.val &&
           isMirrorDFS(t1.left, t2.right) &&
           isMirrorDFS(t1.right, t2.left);
}

function isSymmetricDFS(root) {
    if (!root) return true;
    return isMirrorDFS(root.left, root.right);
}

// BFS Approach
function isSymmetricBFS(root) {
    if (!root) return true;

    let queue = [[root.left, root.right]];

    while (queue.length > 0) {
        let [t1, t2] = queue.shift();

        if (!t1 && !t2) continue;
        if (!t1 || !t2 || t1.val !== t2.val) return false;

        queue.push([t1.left, t2.right]);
        queue.push([t1.right, t2.left]);
    }
    return true;
}

function main() 
{
    let root = new TreeNode(1);
    root.left = new TreeNode(2);
    root.right = new TreeNode(2);
    root.left.left = new TreeNode(3);
    root.left.right = new TreeNode(4);
    root.right.left = new TreeNode(4);
    root.right.right = new TreeNode(3);

    console.log("DFS Approach:", isSymmetricDFS(root));
    console.log("BFS Approach:", isSymmetricBFS(root));
}

main();

File: 08. Trees/01. Binary Tree/16. Symmetric Tree/main.py
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# DFS Approach
def is_mirror_dfs(t1, t2):
    if not t1 and not t2:
        return True
    if not t1 or not t2:
        return False
    return (t1.val == t2.val and 
            is_mirror_dfs(t1.left, t2.right) and 
            is_mirror_dfs(t1.right, t2.left))

def is_symmetric_dfs(root):
    if not root:
        return True
    return is_mirror_dfs(root.left, root.right)

# BFS Approach
def is_symmetric_bfs(root):
    if not root:
        return True

    queue = deque([(root.left, root.right)])

    while queue:
        t1, t2 = queue.popleft()

        if not t1 and not t2:
            continue
        if not t1 or not t2 or t1.val != t2.val:
            return False

        queue.append((t1.left, t2.right))
        queue.append((t1.right, t2.left))

    return True

# Main function
def main():
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(2)
    root.left.left = TreeNode(3)
    root.left.right = TreeNode(4)
    root.right.left = TreeNode(4)
    root.right.right = TreeNode(3)

    print("DFS Approach:", is_symmetric_dfs(root))
    print("BFS Approach:", is_symmetric_bfs(root))

main()
